src/types.rs

```diff
@@ -825,6 +825,166 @@ impl SourceMap {
 
         Ok((sm, mapping))
     }
+
+    /// Adjusts the mappings in `original` using the mappings in `adjustment`.
+    ///
+    /// Here is the intended use case for this function:
+    /// * You have a source file (for example, minified JS) `foo.js` and a
+    ///   corresponding sourcemap `foo.js.map`.
+    /// * You modify `foo.js` in some way and generate a sourcemap `transform.js.map`
+    ///   representing this modification. This can be done using `magic-string`, for example.
+    /// * You want a sourcemap that is "like" `foo.js.map`, but takes the changes you made to `foo.js` into account.
+    /// Then `SourceMap::adjust_mappings(foo.js.map, transform.js.map)` is the desired sourcemap.
+    ///
+    /// This function assumes that `adjustment` contains no relevant information except for mappings.
+    /// All information about sources and names is copied from `original`.
+    ///
+    /// Note that the resulting sourcemap will be at most as fine-grained as `original.` For example,
+    /// if `original` maps every line/column to `0/0`, then `SourceMap::adjust_mappings(original, adjustment)`
+    /// will not map to anything other than`0/0`, irrespective of how detailed the mappings in `adjustment` are.
+    pub fn adjust_mappings(original: &Self, adjustment: &Self) -> Self {
+        // The algorithm works by going through the tokens in `original` in order and adjusting
+        // them depending on the token in `adjustment` they're "covered" by.
+        // For example:
+        // Let `l` be a token in `original` mapping `(17, 23)` to `(8, 30)` and let
+        // `r₁ : (8, 28) -> (102, 35)`, `r₂ : (8, 40) -> (102, 50)`, and
+        // `r₃ : (9, 10) -> (103, 12)` be the tokens in `original` that fall in the range of `l`.
+        // `l` offsets these tokens by `(+9, -7)`, so `r₁, … , r₃` must be offset by the same
+        // amount. Thus, the adjusted sourcemap will contain the tokens
+        // `c₁ : (17, 23) -> (102, 35)`, `c₂ : (17, 33) -> (102, 50)`, and
+        // `c3 : (18, 3) -> (103, 12)`.
+        //
+        // Or, in diagram form:
+        //
+        //    (17, 23)                                    (position in the edited source file)
+        //    ↓ l
+        //    (8, 30)
+        // (8, 28)        (8, 40)        (9, 10)          (positions in the original source file)
+        // ↓ r₁           ↓ r₂           ↓ r₃
+        // (102, 35)      (102, 50)      (103, 12)        (positions in the target file)
+        //
+        // becomes
+        //
+        //    (17, 23)       (17, 33)       (18, 3)       (positions in the edited source file)
+        //    ↓ c₁           ↓ c₂           ↓ c₃
+        //    (102, 35)      (102, 50)      (103, 12)     (positions in the target file)
```

Comment:
not really!
tokens on the same line need the column offset, all other tokens on following lines do not.

so c3 should have no changes to its column offset.

Reply:
I tested it, it's correct the way it is.